clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -370,6 +378,200 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
+        int sizeOfBodyInBytes = ByteUtils.readVarint(input);
+        int totalSizeInBytes = ByteUtils.sizeOfVarint(sizeOfBodyInBytes) + sizeOfBodyInBytes;
+
+        return readPartiallyFrom(input, totalSizeInBytes, sizeOfBodyInBytes, baseOffset, baseTimestamp,
+            baseSequence, logAppendTime);
+    }
+
+    private static PartialDefaultRecord readPartiallyFrom(DataInput input,
+                                                          int sizeInBytes,
+                                                          int sizeOfBodyInBytes,
+                                                          long baseOffset,
+                                                          long baseTimestamp,
+                                                          int baseSequence,
+                                                          Long logAppendTime) throws IOException {
+        try {
+            // reading the attributes / timestamp / offset and key-size does not require
+            // any byte array allocation and therefore we can just read them straight-forwardly
+            byte attributes = input.readByte();
+            int bytesRead = 1;
+            long timestampDelta = ByteUtils.readVarlong(input);
+            long timestamp = baseTimestamp + timestampDelta;
+            if (logAppendTime != null)
+                timestamp = logAppendTime;
+            bytesRead += ByteUtils.sizeOfVarlong(timestampDelta);
+
+            int offsetDelta = ByteUtils.readVarint(input);
+            bytesRead += ByteUtils.sizeOfVarint(offsetDelta);
+            long offset = baseOffset + offsetDelta;
+            int sequence = baseSequence >= 0 ?
+                DefaultRecordBatch.incrementSequence(baseSequence, offsetDelta) :
+                RecordBatch.NO_SEQUENCE;
+
+            // now follow a while loop state-machine to read / skip the rest of fields
+            ReadState state = ReadState.READ_KEY_SIZE;
+            int bytesRemaining = sizeOfBodyInBytes - bytesRead;
+
+            int size = Math.min(MAX_SKIP_BUFFER_SIZE, bytesRemaining);
+            byte[] array = new byte[size];
+            ByteBuffer buffer = ByteBuffer.wrap(array);
+            // set its limit to 0 to indicate no bytes readable yet
+            buffer.limit(0);
+
+            boolean notDone = true;
+            boolean needMore = true;
+            int keySize = -1;
+            int valueSize = -1;
+            int remainingKeyBytesToSkip = -1;
+            int remainingValueBytesToSkip = -1;
+            int remainingHeadersToSkip = -1;
+
+            int recordKeySize = -1;
+            int recordValueSize = -1;
+
+            while (notDone) {
+                if (needMore) {
+                    if (bytesRemaining > 0) {
+                        // first copy the remaining bytes to the beginning of the array
+                        // do not use System.arrayCopy since it will allocate a new array for same src/dest
```

Comment:
Yes but I cannot say for sure if they would not allocate a temporary array or not: if src.offset > dst.offst then it is quite straight-forward (as in our case), but if src.offset < dst.offset then it's a bit tricky how to do that right without a temp array. I just wanted to be on the safer side since it is a quite critical code path. WDYT @hachikuji ?

Reply:
Safer side in what way? If it's a performance thing, then you have to measure. `arraycopy` is usually fastest for what it's worth.

Reply:
In this: `System.arrayCopy` relies on its OS for native implementation; even measuring it against one system does not exclude its risk of being ill-performant on other OS with src/dst the same array. I did some search for arrayCopy on how it handles the same src/dst and I cannot confidently say that across all common environments its implementation would be no worse than non-native implementation. Also have seen some JIRA tickets like (https://bugs.openjdk.java.net/browse/JDK-8179678) makes me feel it's safer just do this copying directly, since it usually will only have a few bytes to copy.

Reply:
Agreed that it is better to measure.

Reply:
It doesn't rely on the OS, it's a JVM intrinsic (this is a critical distinction). And it's used all over the place by the collection libraries. We only really care about Linux performance. If it's a small number of bytes, it may not matter, but generally I think perf driven changes have to be measured.

Reply:
Also, there are a number of crashed did to misoptimized loops too (Lucene filed a number of these over time, it doesn't mean we can't use loops).

Reply:
Okay let me speak out one more time: I was wrong saying its `OS` dependent, and it is JDK dependent (e.g. your version of JDK, whether it's openJDK v.s. OracleJDK). But still my point is that, doing one measurement on a single platform does not give us much confidence. On the other hand, although it is a critical code path, we are only moving at most 4 bytes in this function, and we are in the worst case moving 3 + num.headers times (key-size, value-size, num-headers for varint) for each record, which makes feel not worth doing system.arrayCopy and carefully measure its perf under different platform environment, or whether it is error-prone on some JDK versions.

About the comment itself: I can remove it as @hachikuji said it is misleading, but still using arrayCopy here sounds like an over-optimization to me.

Reply:
It's not any more JDK dependent than running a loop is JDK dependent (JIT optimizations vary more than this method that changes less often). Anyway, if it's 4 bytes, it doesn't matter.